### PR TITLE
Fix IE 11 handshake split

### PR DESCRIPTION
**Pull request by BrennanConroy**
Part of investigating sauce lab flakiness

No tests because there is no way to purposefully hit this code path in functional tests.

Tested manually by adding a delay in server code to force this to happen.

### CODE REVIEW

File: src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts [hunk 1]
```diff
@@ -40,7 +40,20 @@ export class HandshakeProtocol {
...
+                // which returns an Array object, that doesn't have .buffer, which is an ArrayBuffer
+                // so we need to convert it into a Uint8Array and .buffer that
+                if (remainingData.buffer) {
+                    remainingData = remainingData.buffer;
+                } else {
+                    remainingData = new Uint8Array(remainingData).buffer;
+                }
+            } else {
+                remainingData = null;
+            }
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts]: Does this mean we can change the signature from `(data: any): [any, HandshakeResponseMessage]` to `(data: any): [ArrayBuffer, HandshakeResponseMessage]`?

Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts]: We could probably change it to `(data: any): [ArrayBuffer | String, HandshakeResponseMessage]`

File: src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts [hunk 2]
```diff
@@ -40,7 +40,20 @@ export class HandshakeProtocol {
             // optional content after is additional messages
             const responseLength = separatorIndex + 1;
             messageData = String.fromCharCode.apply(null, binaryData.slice(0, responseLength));
-            remainingData = (binaryData.byteLength > responseLength) ? binaryData.slice(responseLength).buffer : null;
+
+            if (binaryData.byteLength > responseLength) {
+                remainingData = binaryData.slice(responseLength);
+                // IE 11 does not have .slice() on Uint8Array, we polyfill it with Array.prototype.slice()
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts]: Wouldn't it be better to just change the polyfill in browser-index.ts to convert the return value back to a Uint8Array?

File: src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts [hunk 3]
```diff
@@ -40,7 +40,20 @@ export class HandshakeProtocol {
...
+                // which returns an Array object, that doesn't have .buffer, which is an ArrayBuffer
+                // so we need to convert it into a Uint8Array and .buffer that
+                if (remainingData.buffer) {
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts]: If we don't go with the polyfill change, it seems better to create an `isUint8Array` method here and use that.

File: src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts (hunk 2 above)
Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/HandshakeProtocol.ts]: I like that idea better :D

**Review by halter73 — APPROVED** [review]

**Review by mikaelm12 — APPROVED** [review]
Weird and Javascripty

**Pull request merged**